File: app/src/main/java/com/battlelancer/seriesguide/ui/shows/NowAdapter.java
package com.battlelancer.seriesguide.ui.shows;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.thetvdbapi.TvdbImageTools;
import com.battlelancer.seriesguide.util.ServiceUtils;
import com.battlelancer.seriesguide.util.TextTools;
import com.battlelancer.seriesguide.util.TimeTools;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * Sectioned adapter displaying recently watched episodes and episodes recently watched by trakt
 * friends.
 */
public class NowAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    protected static final String TRAKT_ACTION_WATCH = "watch";

    public interface ItemClickListener {
        void onItemClick(View view, int position);
    }

    protected static class HistoryViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.textViewHistoryShow) public TextView show;
        @BindView(R.id.textViewHistoryEpisode) public TextView episode;
        @BindView(R.id.imageViewHistoryPoster) public ImageView poster;
        @BindView(R.id.textViewHistoryInfo) public TextView info;
        @BindView(R.id.imageViewHistoryAvatar) public ImageView avatar;
        @BindView(R.id.imageViewHistoryType) public ImageView type;

        public HistoryViewHolder(View itemView, final ItemClickListener listener) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (position != RecyclerView.NO_POSITION && listener != null) {
                    listener.onItemClick(v, position);
                }
            });
        }
    }

    static class MoreViewHolder extends RecyclerView.ViewHolder {
        public TextView title;

        public MoreViewHolder(View itemView, final ItemClickListener listener) {
            super(itemView);
            title = itemView.findViewById(R.id.textViewNowMoreText);

            itemView.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (position != RecyclerView.NO_POSITION && listener != null) {
                    listener.onItemClick(v, position);
                }
            });
        }
    }

    static class HeaderViewHolder extends RecyclerView.ViewHolder {
        public TextView title;

        public HeaderViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.textViewGridHeader);
        }
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({ItemType.HISTORY, ItemType.FRIEND, ItemType.MORE_LINK, ItemType.HEADER})
    public @interface ItemType {
        int HISTORY = 1;
        int FRIEND = 2;
        int MORE_LINK = 3;
        int HEADER = 4;
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({ViewType.HISTORY, ViewType.MORE_LINK, ViewType.HEADER})
    public @interface ViewType {
        int HISTORY = 1;
        int MORE_LINK = 2;
        int HEADER = 3;
    }

    private final Context context;
    private final ItemClickListener listener;
    private final Drawable drawableWatched;
    private final Drawable drawableCheckin;

    private List<NowItem> dataset;
    private List<NowItem> recentlyWatched;
    private List<NowItem> friendsRecently;

    public static class NowItem {
        public Integer episodeTvdbId;
        public Integer showTvdbId;
        public Integer movieTmdbId;
        public long timestamp;
        public String title;
        public String description;
        public String network;
        public String tvdbPosterUrl;
        public String username;
        public String avatar;
        public String action;
        @ItemType public int type;

        public NowItem recentlyWatchedLocal() {
            this.type = ItemType.HISTORY;
            return this;
        }

        public NowItem recentlyWatchedTrakt(@Nullable String action) {
            this.action = action;
            this.type = ItemType.HISTORY;
            return this;
        }

        public NowItem friend(String username, String avatar, String action) {
            this.username = username;
            this.avatar = avatar;
            this.action = action;
            this.type = ItemType.FRIEND;
            return this;
        }

        public NowItem tvdbIds(Integer episodeTvdbId, Integer showTvdbId) {
            this.episodeTvdbId = episodeTvdbId;
            this.showTvdbId = showTvdbId;
            return this;
        }

        public NowItem tmdbId(Integer movieTmdbId) {
            this.movieTmdbId = movieTmdbId;
            return this;
        }

        public NowItem displayData(long timestamp, String title, String description,
                String tvdbPosterUrl) {
            this.timestamp = timestamp;
            this.title = title;
            this.description = description;
            this.tvdbPosterUrl = tvdbPosterUrl;
            return this;
        }

        public NowItem moreLink(String title) {
            this.type = ItemType.MORE_LINK;
            this.title = title;
            return this;
        }

        public NowItem header(String title) {
            this.type = ItemType.HEADER;
            this.title = title;
            return this;
        }
    }

    public NowAdapter(Context context, ItemClickListener listener) {
        this.context = context;
        this.listener = listener;
        this.dataset = new ArrayList<>();
        this.drawableWatched = AppCompatResources.getDrawable(getContext(),
                R.drawable.ic_watch_16dp);
        this.drawableCheckin = AppCompatResources.getDrawable(getContext(),
                R.drawable.ic_checkin_16dp);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        if (viewType == ViewType.HEADER) {
            View v = LayoutInflater.from(viewGroup.getContext())
                    .inflate(R.layout.item_grid_header, viewGroup, false);
            return new HeaderViewHolder(v);
        } else if (viewType == ViewType.MORE_LINK) {
            View v = LayoutInflater.from(viewGroup.getContext())
                    .inflate(R.layout.item_now_more, viewGroup, false);
            return new MoreViewHolder(v, listener);
        } else if (viewType == ViewType.HISTORY) {
            return getHistoryViewHolder(viewGroup, listener);
        } else {
            throw new IllegalArgumentException("Using unrecognized view type.");
        }
    }

    @NonNull
    protected RecyclerView.ViewHolder getHistoryViewHolder(ViewGroup viewGroup,
            ItemClickListener itemClickListener) {
        View v = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.item_history, viewGroup, false);
        return new HistoryViewHolder(v, itemClickListener);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int position) {
        NowItem item = getItem(position);

        if (viewHolder instanceof HeaderViewHolder) {
            HeaderViewHolder holder = (HeaderViewHolder) viewHolder;

            holder.title.setText(item.title);
        } else if (viewHolder instanceof MoreViewHolder) {
            MoreViewHolder holder = (MoreViewHolder) viewHolder;

            holder.title.setText(item.title);
        } else if (viewHolder instanceof HistoryViewHolder) {
            HistoryViewHolder holder = (HistoryViewHolder) viewHolder;

            String time = TimeTools.formatToLocalRelativeTime(getContext(),
                    new Date(item.timestamp));
            if (item.type == ItemType.HISTORY) {
                // user history entry
                holder.avatar.setVisibility(View.GONE);
                holder.info.setText(time);
            } else {
                // friend history entry
                holder.avatar.setVisibility(View.VISIBLE);
                holder.info.setText(TextTools.dotSeparate(item.username, time));

                // trakt avatar
                ServiceUtils.loadWithPicasso(getContext(), item.avatar).into(holder.avatar);
            }

            // a TVDb or no poster
            TvdbImageTools.loadShowPosterResizeSmallCrop(getContext(), holder.poster,
                    item.tvdbPosterUrl);

            holder.show.setText(item.title);
            holder.episode.setText(item.description);

            // action type indicator (only if showing trakt history)
            if (TRAKT_ACTION_WATCH.equals(item.action)) {
                holder.type.setImageDrawable(getDrawableWatched());
                holder.type.setEnabled(false);
            } else if (item.action != null) {
                // check-in, scrobble
                holder.type.setImageDrawable(getDrawableCheckin());
                holder.type.setVisibility(View.VISIBLE);
            } else {
                holder.type.setVisibility(View.GONE);
            }
            // Set disabled for darker icon (non-interactive).
            holder.type.setEnabled(false);
        }
    }

    @Override
    public int getItemCount() {
        return dataset.size();
    }

    @Override
    public int getItemViewType(int position) {
        int itemType = getItem(position).type;
        switch (itemType) {
            case ItemType.HISTORY:
            case ItemType.FRIEND:
                return ViewType.HISTORY;
            case ItemType.MORE_LINK:
                return ViewType.MORE_LINK;
            case ItemType.HEADER:
                return ViewType.HEADER;
        }
        return 0;
    }

    protected Context getContext() {
        return context;
    }

    protected Drawable getDrawableWatched() {
        return drawableWatched;
    }

    protected Drawable getDrawableCheckin() {
        return drawableCheckin;
    }

    public NowItem getItem(int position) {
        return dataset.get(position);
    }

    public void setRecentlyWatched(List<NowItem> items) {
        int oldCount = recentlyWatched == null ? 0 : recentlyWatched.size();
        int newCount = items == null ? 0 : items.size();

        recentlyWatched = items;
        reloadData();
        notifyAboutChanges(0, oldCount, newCount);
    }

    public void setFriendsRecentlyWatched(List<NowItem> items) {
        int oldCount = friendsRecently == null ? 0 : friendsRecently.size();
        int newCount = items == null ? 0 : items.size();
        // items start after recently watched (if any)
        int startPosition = recentlyWatched == null ? 0 : recentlyWatched.size();

        friendsRecently = items;
        reloadData();
        notifyAboutChanges(startPosition, oldCount, newCount);
    }

    private void reloadData() {
        dataset.clear();
        if (recentlyWatched != null) {
            dataset.addAll(recentlyWatched);
        }
        if (friendsRecently != null) {
            dataset.addAll(friendsRecently);
        }
    }

    private void notifyAboutChanges(int startPosition, int oldItemCount, int newItemCount) {
        if (newItemCount == 0 && oldItemCount == 0) {
            return;
        }

        if (newItemCount == oldItemCount) {
            // identical number of items
            notifyItemRangeChanged(startPosition, oldItemCount);
        } else if (newItemCount > oldItemCount) {
            // more items than before
            if (oldItemCount > 0) {
                notifyItemRangeChanged(startPosition, oldItemCount);
            }
            notifyItemRangeInserted(startPosition + oldItemCount,
                    newItemCount - oldItemCount);
        } else {
            // less items than before
            if (newItemCount > 0) {
                notifyItemRangeChanged(startPosition, newItemCount);
            }
            notifyItemRangeRemoved(startPosition + newItemCount,
                    oldItemCount - newItemCount);
        }
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/history/NowAdapter.kt
package com.battlelancer.seriesguide.shows.history

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.IntDef
import androidx.appcompat.content.res.AppCompatResources
import androidx.recyclerview.widget.RecyclerView
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.ItemHistoryBinding

/**
 * Sectioned adapter displaying recently watched episodes and episodes recently watched by
 * friends.
 */
open class NowAdapter(
    protected val context: Context,
    private val listener: ItemClickListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    interface ItemClickListener {
        fun onItemClick(view: View, position: Int)
    }

    internal class MoreViewHolder(itemView: View, listener: ItemClickListener) :
        RecyclerView.ViewHolder(itemView) {
        var title: TextView

        init {
            title = itemView.findViewById(R.id.textViewNowMoreText)
            itemView.setOnClickListener { v: View ->
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    listener.onItemClick(v, position)
                }
            }
        }
    }

    internal class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var title: TextView

        init {
            title = itemView.findViewById(R.id.textViewGridHeader)
        }
    }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(ItemType.HISTORY, ItemType.FRIEND, ItemType.MORE_LINK, ItemType.HEADER)
    annotation class ItemType {
        companion object {
            const val HISTORY = 1
            const val FRIEND = 2
            const val MORE_LINK = 3
            const val HEADER = 4
        }
    }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(ViewType.HISTORY, ViewType.MORE_LINK, ViewType.HEADER)
    annotation class ViewType {
        companion object {
            const val HISTORY = 1
            const val MORE_LINK = 2
            const val HEADER = 3
        }
    }

    class NowItem {
        var episodeRowId: Long? = null
        var showTmdbId: Int? = null
        var movieTmdbId: Int? = null
        var timestamp: Long = 0
        var title: String? = null
        var description: String? = null
        var network: String? = null
        var posterUrl: String? = null
        var username: String? = null
        var avatar: String? = null
        var action: String? = null

        @ItemType
        var type = 0

        fun recentlyWatchedLocal(): NowItem {
            this.type = ItemType.HISTORY
            return this
        }

        fun recentlyWatchedTrakt(action: String?): NowItem {
            this.action = action
            this.type = ItemType.HISTORY
            return this
        }

        fun friend(username: String?, avatar: String?, action: String?): NowItem {
            this.username = username
            this.avatar = avatar
            this.action = action
            this.type = ItemType.FRIEND
            return this
        }

        /**
         * Pass 0 if no value.
         */
        fun episodeIds(episodeRowId: Long, showTmdbId: Int): NowItem {
            this.episodeRowId = episodeRowId
            this.showTmdbId = showTmdbId
            return this
        }

        fun tmdbId(movieTmdbId: Int?): NowItem {
            this.movieTmdbId = movieTmdbId
            return this
        }

        fun displayData(
            timestamp: Long, title: String?, description: String?,
            posterUrl: String?
        ): NowItem {
            this.timestamp = timestamp
            this.title = title
            this.description = description
            this.posterUrl = posterUrl
            return this
        }

        fun moreLink(title: String): NowItem {
            this.type = ItemType.MORE_LINK
            this.title = title
            return this
        }

        fun header(title: String): NowItem {
            this.type = ItemType.HEADER
            this.title = title
            return this
        }
    }

    protected val drawableWatched: Drawable
    protected val drawableCheckin: Drawable

    private val dataset: MutableList<NowItem>
    private var recentlyWatched: List<NowItem>? = null
    private var friendsRecently: List<NowItem>? = null

    init {
        dataset = ArrayList()
        drawableWatched = AppCompatResources.getDrawable(context, R.drawable.ic_watch_16dp)!!
        drawableCheckin = AppCompatResources.getDrawable(context, R.drawable.ic_checkin_16dp)!!
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            ViewType.HEADER -> {
                val v = LayoutInflater.from(viewGroup.context)
                    .inflate(R.layout.item_grid_header, viewGroup, false)
                HeaderViewHolder(v)
            }
            ViewType.MORE_LINK -> {
                val v = LayoutInflater.from(viewGroup.context)
                    .inflate(R.layout.item_now_more, viewGroup, false)
                MoreViewHolder(v, listener)
            }
            ViewType.HISTORY -> {
                getHistoryViewHolder(viewGroup, listener)
            }
            else -> {
                throw IllegalArgumentException("Using unrecognized view type.")
            }
        }
    }

    private fun getHistoryViewHolder(
        viewGroup: ViewGroup,
        itemClickListener: ItemClickListener
    ): RecyclerView.ViewHolder {
        return HistoryViewHolder(
            ItemHistoryBinding.inflate(
                LayoutInflater.from(viewGroup.context), viewGroup, false
            ),
            itemClickListener
        )
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        val item = getItem(position)
        when (viewHolder) {
            is HeaderViewHolder -> {
                viewHolder.title.text = item.title
            }
            is MoreViewHolder -> {
                viewHolder.title.text = item.title
            }
            is HistoryViewHolder -> {
                viewHolder.bindToShow(context, item, drawableWatched, drawableCheckin)
            }
        }
    }

    override fun getItemCount(): Int = dataset.size

    override fun getItemViewType(position: Int): Int {
        return when (getItem(position).type) {
            ItemType.HISTORY, ItemType.FRIEND -> ViewType.HISTORY
            ItemType.MORE_LINK -> ViewType.MORE_LINK
            ItemType.HEADER -> ViewType.HEADER
            else -> 0
        }
    }

    fun getItem(position: Int): NowItem = dataset[position]

    fun setRecentlyWatched(items: List<NowItem>?) {
        val oldCount = recentlyWatched?.size ?: 0
        val newCount = items?.size ?: 0
        recentlyWatched = items
        reloadData()
        notifyAboutChanges(0, oldCount, newCount)
    }

    fun setFriendsRecentlyWatched(items: List<NowItem>?) {
        val oldCount = friendsRecently?.size ?: 0
        val newCount = items?.size ?: 0
        // items start after recently watched (if any)
        val startPosition = recentlyWatched?.size ?: 0
        friendsRecently = items
        reloadData()
        notifyAboutChanges(startPosition, oldCount, newCount)
    }

    private fun reloadData() {
        dataset.clear()
        val recentlyWatched = recentlyWatched
        if (recentlyWatched != null) {
            dataset.addAll(recentlyWatched)
        }
        val friendsRecently = friendsRecently
        if (friendsRecently != null) {
            dataset.addAll(friendsRecently)
        }
    }

    private fun notifyAboutChanges(startPosition: Int, oldItemCount: Int, newItemCount: Int) {
        if (newItemCount == 0 && oldItemCount == 0) {
            return
        }
        if (newItemCount == oldItemCount) {
            // identical number of items
            notifyItemRangeChanged(startPosition, oldItemCount)
        } else if (newItemCount > oldItemCount) {
            // more items than before
            if (oldItemCount > 0) {
                notifyItemRangeChanged(startPosition, oldItemCount)
            }
            notifyItemRangeInserted(
                startPosition + oldItemCount,
                newItemCount - oldItemCount
            )
        } else {
            // less items than before
            if (newItemCount > 0) {
                notifyItemRangeChanged(startPosition, newItemCount)
            }
            notifyItemRangeRemoved(
                startPosition + newItemCount,
                oldItemCount - newItemCount
            )
        }
    }

    companion object {
        const val TRAKT_ACTION_WATCH = "watch"
    }
}